core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -863,6 +870,29 @@ private[spark] class TaskSchedulerImpl(
       executorUpdates)
   }
 
+ private def getTaskAccumulableInfosAndProgressRate(
+      updates: Seq[AccumulatorV2[_, _]]): (Seq[AccumulableInfo], Double) = {
+   var records = 0L
+   var runTime = 0L
+   val accInfos = updates.map { acc =>
+     if (calculateTaskProgressRate && acc.name.isDefined) {
+       val name = acc.name.get
+       if (name == shuffleRead.RECORDS_READ || name == input.RECORDS_READ) {
+         records += acc.value.asInstanceOf[Long]
+       } else if (name == InternalAccumulator.EXECUTOR_RUN_TIME) {
+         runTime = acc.value.asInstanceOf[Long]
+       }
+     }
+     acc.toInfo(Some(acc.value), None)
+   }
+   val taskProgressRate = if (calculateTaskProgressRate && runTime > 0) {
+     records / (runTime / 1000.0)
+   } else {
+     0.0D
+   }
```

Comment:
Can we centralize the calculation of task progress rate to the `InefficientTaskCalculator` only? It seems not each calculation is necessary here since the speculation check only happens under certain conditions, e.g., `numSuccessfulTasks >= minFinishedForSpeculation`.  And I think we can reuse the existing`TaskInfo._accumulables` directly, which could make cold cleaner.

Reply:
If we calculate  the progress task's taskProgressRate in InefficientTaskCalculator, we  can only  use the TaskInfo. _accumulables insead of  accumUpdates,  while  the TaskInfo._accumulables is updated by event which may be lost and the TaskInfo._accumulables may unreliable.

With the @mridulm suggestions  on https://github.com/apache/spark/pull/36162#discussion_r865591651
For in progress tasks: we can update it  via executor heartbeat , it is not only  latest but also reliable. Getting the taskProgressRate  is based on existing traversal logic,  additional calculations can be ignored.

Reply:
OK..so could we just set `accumUpdates` to the `TaskInfo` here and calculate the rate in `InefficientTaskCalculator` when it's required?

Reply:
Setting the accumUpdates to the TaskInfo may be unnecessary, there are two reasons:
1. We only need the records and runTime in accumUpdates
2. Setting the accumUpdates to the TaskInfo may  take up more storage space, and  calculating the rate also should  traverse it.

With your suggestions ,  we may only set  the records and runTime to the TaskInfo and  calculate the rate in InefficientTaskCalculator when it's required.  But, if we do that , we should make sure that setting and reading the records and runTime with lock,  which  may make the logic more complicated than that we only set a  taskProgressRate to the TaskInfo, right ?